gh-119349: Add ctypes.util function to list loaded shared libraries

WardBrian opened the pull request:
The `dllist()` function calls platform-specific APIs in order to list the runtime libraries loaded by Python and any imported modules. If these fail for any reason, None is returned.


The name could surely be bikeshedded, at the moment it is taken from [the Julia function of the same name](https://docs.julialang.org/en/v1/stdlib/Libdl/#Base.Libc.Libdl.dllist)

Previous discussions:
- https://github.com/python/cpython/issues/119349
- https://discuss.python.org/t/a-ctypes-function-to-list-all-loaded-shared-libraries/36370

Code based on https://github.com/WardBrian/dllist by myself with a contribution from @eryksun 

<!-- gh-issue-number: gh-119349 -->
* Issue: gh-119349
<!-- /gh-issue-number -->


<!-- readthedocs-preview cpython-previews start -->
----
📚 Documentation preview 📚: https://cpython-previews--122946.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

WardBrian committed:
Add function to list the currently loaded libraries to ctypes.util

The dllist() function calls platform-specific APIs in order to
list the runtime libraries loaded by Python and any imported modules.
If these fail for any reason, None is returned.

ZeroIntensity commented on Doc/library/ctypes.rst:
```diff
@@ -2007,6 +2033,18 @@ Utility functions
+   returns ``None``.
+
+   The exact functionality is system dependent.
+
```
```suggestion
   .. versionadded:: 3.14
```

ZeroIntensity commented on Lib/ctypes/util.py:
```diff
@@ -67,6 +67,75 @@ def find_library(name):
+        wintypes.LPDWORD,
+    )
+
+    def _get_module_filename(hModule: wintypes.HMODULE):
```
Do we allow `ctypes` to avoid PEP 8 naming for Windows names? I haven't seen any precedent for this in the stdlib

WardBrian replied:
Renamed these variables

ZeroIntensity commented on Lib/ctypes/util.py:
```diff
@@ -341,6 +448,61 @@ def find_library(name):
+# https://man.openbsd.org/dl_iterate_phdr
+# https://docs.oracle.com/cd/E88353_01/html/E37843/dl-iterate-phdr-3c.html
+# this relies on find_library, which is why it is defined at the end
+if os.name == "posix" and not sys.platform.startswith("aix") and sys.platform not in {"darwin", "ios", "tvos", "watchos"}:
```
Could we split up this line a little bit? This is too long for PEP 8
```suggestion
if os.name == "posix" and not sys.platform.startswith("aix") and sys.platform not in {"darwin", "ios", "tvos", "watchos"}:
```

ZeroIntensity commented on Lib/test/test_ctypes/test_dllist.py:
```diff
@@ -0,0 +1,72 @@
+
+
+WINDOWS = os.name == 'nt'
+APPLE = sys.platform in {"darwin", "ios", "tvos", "watchos"}
```
It's worth noting that Python also supports Android, see [PEP 738](https://peps.python.org/pep-0738). I don't know all that much about Android -- do we need an extra case for the system library, or is it just `libc.so`?

WardBrian replied:
I am also not deeply familiar with android, but it appears it is `libc.so` ([stackoverflow](https://stackoverflow.com/questions/35329621/where-are-all-the-shared-libraries-stored-on-android))

ZeroIntensity commented on Lib/test/test_ctypes/test_dllist.py:
```diff
@@ -0,0 +1,72 @@
+elif AIX:
+    SYSTEM_LIBRARY = None
+else:
+    SYSTEM_LIBRARY = 'libc.so'
```
Is there a performance benefit to this? It might suffice to just change this to:

```suggestion
SYSTEM_LIBRARY = find_library("c")
```

WardBrian replied:
I think it is a bit more self-documenting to list out a library we know we expect on each platform. Some sort of conditional is required in order to handle unsupported platforms (AIX), or Windows, since:

> On Windows, [find_library()](https://docs.python.org/3/library/ctypes.html#ctypes.util.find_library) searches along the system search path, and returns the full pathname, but since there is no predefined naming scheme a call like find_library("c") will fail and return None.

WardBrian committed:
PEP-8 compilance

WardBrian committed:
Detect availability at runtime on non-Apple POSIX

picnixz commented on Lib/ctypes/util.py:
```diff
@@ -67,6 +67,75 @@ def find_library(name):
+    # https://learn.microsoft.com/windows/win32/api/psapi/nf-psapi-enumprocessmodules
+    # https://learn.microsoft.com/windows/win32/api/libloaderapi/nf-libloaderapi-getmodulefilenamew
+
+    _kernel32 = ctypes.WinDLL('kernel32', use_last_error=True)
```
Wouldn't it automatically load a `WinDLL` upon import?

WardBrian replied:
This does actually instantiate the WinDLL in global scope. I could move it inside the function at the cost of it being re-loaded each call, or made this lazy-loaded, if desired

eryksun replied:
"kernel32.dll" is always loaded at process startup, so this just increments the module reference count. The memory for the `WinDLL` object itself is negligible.

WardBrian replied:
That'd what I'd assumed - the POSIX code unconditionally loads libc, for the same reasons. 

This win32 code also loads `psapi.dll` on the next line, do you think that is worth making lazy, or is it similar?

eryksun replied:
I wouldn't worry about "psapi.dll" either. The loaded module is tiny and loads pretty much instantly. Nowadays it's just stub functions that jump to the real functions in "kernelbase.dll". The latter is loaded at process startup.

WardBrian replied:
Great, thanks!

picnixz commented on Lib/ctypes/util.py:
```diff
@@ -80,6 +149,39 @@ def find_library(name):
+            for i in range(1, num_images):
+                raw_name = libc._dyld_get_image_name(i)
+                try:
+                    name = raw_name.decode("utf-8")
```
Should the names necessarily be UTF-8 encodable? or is there a way to have locale-dependent names? Instead, why not just return a list of bytes? If not, use `os.fsdecode` instead (I *think* the name should be a path, right?)

WardBrian replied:
I think os.fsdecode is the better answer, thanks!

picnixz commented on Lib/ctypes/util.py:
```diff
@@ -80,6 +149,39 @@ def find_library(name):
+                try:
+                    name = raw_name.decode("utf-8")
+                    libraries.append(name)
+                except:
```
This except is way too broad. KeyboardInterrupt should not be caught here IMO. In addition `libraries.append` shouldn't fail.

picnixz commented on Lib/ctypes/util.py:
```diff
@@ -80,6 +149,39 @@ def find_library(name):
+
+            return libraries
+        except Exception as e:
+            import warnings
```
I'd prefer not to have warnings. Other functions don't warn when they can't get their results.

WardBrian replied:
Including when the platform itself is unsupported (in other words, in cases where this function will never return anything but None)?

picnixz replied:
I think so. I think the documentation should just say "if the platform is unsupported or if the DLLs cannot be retrieved, this returns ``None``". But this is my personal opinion and I don't mind a core dev telling me that this or that would be better. I see this function as a utility that may or may not be fiable.

WardBrian replied:
Fair enough. I've removed them for now, they're easy enough to add back if requested. But I agree, this is definitely a 'best-effort' utility function

picnixz commented on Lib/ctypes/util.py:
```diff
@@ -341,6 +443,76 @@ def find_library(name):
+# https://docs.oracle.com/cd/E88353_01/html/E37843/dl-iterate-phdr-3c.html
+# this relies on find_library, which is why it is defined at the end
+if (os.name == "posix" and
+    sys.platform not in {"darwin", "ios", "tvos", "watchos"}):
```
Can't you put this in the `elif os.name == 'posix'`?

WardBrian replied:
That would miss android and AIX (Android, which is supported, and AIX, which is not, but will hit the case where libc doesn't define the required symbol)

picnixz commented on Lib/ctypes/util.py:
```diff
@@ -341,6 +443,76 @@ def find_library(name):
+                    # remove the first entry, which is the executable itself
+                    libraries.pop(0)
+
+                return libraries
```
```suggestion
                # remove the first entry, which is the executable itself
                return libraries[1:]
```

picnixz replied:
(I hope that `libraries` is assured to be a list, now I don't remember)

picnixz commented on Lib/ctypes/util.py:
```diff
@@ -341,6 +443,76 @@ def find_library(name):
+            return None
+
+    else:
+
```
Remove un-necessary empty lines (see the style of the rest of the file).

WardBrian replied:
This file is inconsistent - see e.g. lines 283-285 and lines 380-384

I'm happy to go with whichever is preferred

picnixz replied:
Honestly, I think we should be consistent with the rest of the library. Most of the time, we have *no* empty lines even inside functions. I don't think we should have empty lines around `else` or 2 consecutive empty lines in a function for instance.

WardBrian replied:
Removed

picnixz commented on Lib/ctypes/util.py:
```diff
@@ -341,6 +443,76 @@ def find_library(name):
+        def _info_callback(info, _size, data):
+            libraries = data.contents.value
+            try:
+                name = info.contents.dlpi_name.decode("utf-8")
```
`os.fsdecode` here (if my assumption is correct)

picnixz commented on Lib/ctypes/util.py:
```diff
@@ -341,6 +443,76 @@ def find_library(name):
+            try:
+                name = info.contents.dlpi_name.decode("utf-8")
+                libraries.append(name)
+            except:
```
try to narrow the exception types being caught

picnixz commented on Lib/test/test_ctypes/test_dllist.py:
```diff
@@ -0,0 +1,72 @@
+        dlls = dllist()
+
+        if dlls is not None:
+            from test.support import import_helper
```
You can import it at the top.

picnixz commented on Lib/test/test_ctypes/test_dllist.py:
```diff
@@ -0,0 +1,72 @@
+            from test.support import import_helper
+            _ctypes_test = import_helper.import_module("_ctypes_test")
+            test_module = CDLL(_ctypes_test.__file__)
+
```
Ditto for empty lines

picnixz commented on Misc/NEWS.d/next/Library/2024-08-12-11-58-15.gh-issue-119349.-xTnHl.rst:
```diff
@@ -0,0 +1,2 @@
+Added the :func:`ctypes.util.dllist` function to list the loaded shared
```
This also requires a What's New entry (it's more than just a small entry I'd say)

WardBrian replied:
Added

picnixz commented on Lib/test/test_ctypes/test_dllist.py:
```diff
@@ -0,0 +1,72 @@
+                # The test module is already loaded
+                if test.support.verbose:
+                    print("The test module is already loaded")
+                return
```
You can do a `skipTest` instead.

picnixz commented on Lib/test/test_ctypes/test_dllist.py:
```diff
@@ -0,0 +1,72 @@
+                for dll in dlls:
+                    print("\t", dll)
+            else:
+                print("\tFunction returned None")
```
Instead of this, you could add a message to display when your test fails (e.g., `self.assertTrue(..., f'loaded={dlls}')`).

picnixz commented on Lib/test/test_ctypes/test_dllist.py:
```diff
@@ -0,0 +1,72 @@
+                    print("\t", dll)
+
+            self.assertIsNotNone(dlls2)
+            self.assertGreater(len(dlls2), len(dlls))
```
I think you can refine the test. IIUC, you want `dlls2` to be a superset of `dlls` right? you can do `self.assertGreater(set(dlls2), set(dlls))` for instance.

picnixz commented on Lib/ctypes/util.py:
```diff
@@ -67,6 +67,75 @@ def find_library(name):
+            if not _psapi.EnumProcessModules(process,
+                                            modules,
+                                            ctypes.sizeof(modules),
+                                            ctypes.byref(space_needed)):
```
```suggestion
            if not _psapi.EnumProcessModules(process,
                                             modules,
                                             ctypes.sizeof(modules),
                                             ctypes.byref(space_needed)):
```

picnixz reviewed (commented):
I'm not a fan of emitting warnings. Either you silently skip the failed names, or you raise. Or you return bytes and you won't have decoding issues for instance. But if you can list the DLLs being loaded, you should always return the complete list, and if you can't, then you should return `None` (other functions seem to do it as well when they can't get the result for whatever reason).

WardBrian committed:
use os.fsdecode for library paths

WardBrian committed:
Test improvements

WardBrian committed:
Add What's New entry

WardBrian committed:
Remove warning code

WardBrian committed:
Lint fix

WardBrian committed:
Avoid editing cached versions of symbols

ZeroIntensity commented on Lib/ctypes/util.py:
```diff
@@ -80,6 +141,28 @@ def find_library(name):
+
+            return libraries
+        except Exception:
+            return None
```
I still think this is bad design, we should let the error propagate. As you said, there's not much a user could do if an exception occurs, but that goes for a `None` return as well.

FWIW, if we do go the route of returning `None` (although, that would take convincing), catching `Exception` is the wrong way to do it here. If something unrelated goes wrong (e.g. out of memory, some internal corruption, or a million other things that could go wrong), *that* should definitely propagate -- move the `except Exception` to specific lines, and expect specific exceptions.

WardBrian replied:
To me the difference arises if I'm in an application where I want to treat errors in the same way as "couldn't find the library I'm looking for".

If we return None, then I can write 

`'libfoo.so' in (dllist() or [])`

If dllist can throw an exception, this same logic becomes something like 
```python
dlls = []
try:
	dlls = dllist()
except:
	pass

if 'libfoo.so' in dlls:
	...
```
There's no real information gained in the second case, but handling an exception is much more verbose than handling a value for this.

---

Conversely, if we _don't_ want to treat it like this, you can write this in the None case:

```python
dlls = dllist()
if dlls is None:
	raise RuntimeError("could not determine if libfoo was loaded")

if 'libfoo.so' in dlls:
	...
```

Or this, in the other case:

```python
dlls = []
try:
	dlls = dllist()
except Exception as e:
	raise RuntimeError("could not determine if libfoo was loaded") from e

if 'libfoo.so' in dlls:
	...
```

This case is _also_ more verbose, and the only advantage it has is the `from e`, which may or may not contain any meaningful information for the actual user.

---

It's also the same behavior that `ctypes.util.find_library` has, e.g.:
https://github.com/python/cpython/blob/e2f2dc708eae89f41e328501b5ea7c97b8e39907/Lib/ctypes/util.py#L335-L336

ZeroIntensity replied:
My main issue with a `None` return is that you *force* the user to suppress any exception that occurred -- some may want to write to logs, reraise it, or in your case, catch it and assign to an empty list. As you've shown, it's pretty trivial to catch the exception and assign the result to something more useful, but I don't think we should force that choice on everyone.

`find_library` is a bit different to me, because when I think of finding something, it's sort of implied that the thing might not have been found -- *that's* what the `None` return is for, not for indicating that an error occurred.

WardBrian replied:
Would you propose we also raise unconditionally in the "supported platform" case?

ZeroIntensity replied:
I think we could return `None` in some cases, but let errors propagate that are unrelated.

WardBrian replied:
That would make callers check both cases, which seems like a worst-of-both-worlds

WardBrian replied:
I did some more reading on the various APIs that we call, and the only one that can actually "fail" is Windows. MacOS's API might return NULL for the name of some modules, so I'm handling that explicitly, but otherwise we don't need these `except Exception` blocks 

The Windows function only fails in the case
> If the module list in the target process is corrupted or not yet initialized

So while I have left the error handling code there, I do not believe you would ever encounter it while managing to keep Python itself up and running...

ZeroIntensity commented on Lib/ctypes/util.py:
```diff
@@ -67,6 +67,67 @@ def find_library(name):
+                                         ctypes.sizeof(modules),
+                                         ctypes.byref(space_needed)):
+                error = ctypes.get_last_error()
+                raise RuntimeError(f"EnumProcessModules failed with error code {error}")
```
I think this should raise an `OSError` with the error code set.

WardBrian replied:
I used the `ctypes.WinError` subclass of `OSError`. Thanks!

WardBrian committed:
use ctypes.WinError in _get_module_handles

eryksun commented on Lib/ctypes/util.py:
```diff
@@ -67,6 +67,67 @@ def find_library(name):
+                                         ctypes.sizeof(modules),
+                                         ctypes.byref(space_needed)):
+                raise ctypes.WinError(ctypes.get_last_error(),
+                                      "EnumProcessModules failed")
```
```suggestion
                err = ctypes.get_last_error()
                msg = ctypes.FormatError(err).strip()
                raise ctypes.WinError(err, f"EnumProcessModules failed: {msg}")
```

WardBrian committed:
Format WinError message

Co-authored-by: Eryk Sun <eryksun@gmail.com>

WardBrian committed:
Remove unnecessary error handling

ZeroIntensity commented on Lib/ctypes/util.py:
```diff
@@ -67,6 +67,68 @@ def find_library(name):
+            libraries = [name for h in modules
+                            if (name := _get_module_filename(h)) is not None]
+            return libraries
+        except OSError:
```
A nitpick: I don't like `try`ing unrelated code, it should be over just the function call, not the list comp -- errors in getting the file name should propagate.

WardBrian committed:
Tighten except statement

ZeroIntensity commented on Doc/library/ctypes.rst:
```diff
@@ -2007,6 +2033,20 @@ Utility functions
+   Try to provide a list of paths of the shared libraries loaded into the current
+   process.  These paths are not normalized or processed in any way. If the
+   function is not able to determine the list of loaded libraries (either because
+   the current platform does not support it, or because an error occurred), it
```
If the current platform doesn't support it, we should at least document *why*. Could you add an excerpt about the system needing libc to have `dl_iterate_phdr` in order for this to work?

WardBrian replied:
I could, though I could imagine such a thing going stale if specific support for another platform was added down the line, or some other API was used

ZeroIntensity replied:
Sure, and then we document that later on.

WardBrian committed:
Mention reasoning for lack of support in doc

WardBrian committed:
Merge branch 'main' into ctypes-util-list-libraries

WardBrian committed:
Lint fix

WardBrian committed:
Merge branch 'main' into ctypes-util-list-libraries

encukou commented on Doc/library/ctypes.rst:
```diff
@@ -1406,6 +1406,32 @@ the shared library name at development time, and hardcode that into the wrapper
+   the current platform does not support it, or because an error occurred), it
+   returns ``None``.
+
+The exact functionality is system dependent.
```
Let's only have this in one place; users can follow the link.
```suggestion
```

encukou commented on Doc/library/ctypes.rst:
```diff
@@ -2083,6 +2109,21 @@ Utility functions
+   returns ``None``.
+
+   The exact functionality is system dependent.
+
```
Please add `Availability` listing the platforms; ISTM should be
```
.. availability:: Windows, macOS, glibc, BSD libc
```
and perhaps `musl` too. The list is at `Doc/tools/extensions/availability.py`.

WardBrian replied:
I listed those you did. I _believe_ it would also work as-implemented on iOS, Android, and musl, but I have not verified these myself

WardBrian replied:
Testing in an alpine docker image (musl), the code here works _except_ that `find_library('c')` returns none there.
I can still get a handle on libc by using `ctypes.CDLL(None)` -- is this legit to do in the `util` code? 

I see it being used once in the ctypes documentation, but it seems absent from the existing cpython codebase.

encukou replied:
For POSIX, it's fine. Updating the `ctypes.CDLL` docs is tracked in #78773.

encukou commented on Lib/ctypes/util.py:
```diff
@@ -341,6 +418,51 @@ def find_library(name):
+
+        def dllist():
+            libraries = []
+            _libc.dl_iterate_phdr(_info_callback,
```
This should have `argtypes` set.

encukou reviewed (commented):
Please raise an exception if the list can't be retrieved.

WardBrian committed:
Merge branch 'main' into ctypes-util-list-libraries

WardBrian committed:
Raise if list cannot be produced

WardBrian committed:
Stop removing first entry

WardBrian committed:
Set argtypes on dl_iterate_phdr call

WardBrian committed:
Leave dllist undefined if platform doesn't support it

WardBrian committed:
Documentation updates

WardBrian committed:
Move availability above versionadded

WardBrian committed:
Update to work on musl systems

encukou commented on Lib/ctypes/util.py:
```diff
@@ -350,6 +468,7 @@ def test():
         print(cdll.msvcrt)
         print(cdll.load("msvcrt"))
         print(find_library("msvcrt"))
+        print(dllist())
```
Could you instead add something like this for all platforms:
```
    try:
        dllist
    except NameError:
        print('dllist() not available')
    else:
        print(dllist())
```

encukou commented on Doc/library/ctypes.rst:
```diff
@@ -2083,6 +2096,19 @@ Utility functions
+   can raise :exc:`OSError` if the underlying platform APIs fail.
+
+   The first element of the list is typically the path of the current executable
+   file.  The exact functionality is system dependent.
```
```suggestion
   The exact functionality is system dependent.

   On most platforms, the first element of the list represents the current
   executable file. It may be an empty string.
```

WardBrian committed:
Merge branch 'main' into ctypes-util-list-libraries

WardBrian committed:
Update test code

WardBrian committed:
Update documentation

WardBrian committed:
Lint fix

WardBrian committed:
Fix test failing on RHEL8

On that platform, libc is called libc-2.28.so, not libc.so. Instead, the test searches for libffi.so, which is a dependency of ctypes

ZeroIntensity commented on Doc/library/ctypes.rst:
```diff
@@ -2083,6 +2096,20 @@ Utility functions
+   executable file. It may be an empty string.
+
+   .. availability:: Windows, macOS, iOS, glibc, BSD libc, musl
+   .. versionadded:: 3.14
```
We have this cool thing:
```suggestion
   .. versionadded:: next
```

ZeroIntensity commented on Lib/test/test_ctypes/test_dllist.py:
```diff
@@ -0,0 +1,58 @@
+    def test_lists_updates(self):
+        dlls = dllist()
+
+        if dlls is not None:
```
You can remove this now, since it can't return `None`

ZeroIntensity commented on Lib/test/test_ctypes/test_dllist.py:
```diff
@@ -0,0 +1,58 @@
+            dlls1 = set(dlls)
+            dlls2 = set(dlls2)
+            if test.support.verbose:
+                print("Newly loaded shared libraries:")
```
`print` in a test? Preposterous. (Either remove these, or add them to something like a `StringIO` so we can assert the output.)

WardBrian replied:
`print`s guarded by `if test.support.verbose` can also be found in `test_ctypes/test_loading.py` and `test_ctypes/test_find.py`, as well as a small handful of non-ctypes tests. 

I'm happy to remove them if we want, since they were mainly useful for the initial writing of the test

WardBrian replied:
I converted this into part of the assert statements below

encukou replied:
> `print` in a test? Preposterous.

@ZeroIntensity, `print` is OK in verbose mode :)
But putting them in the assert is OK too.

ZeroIntensity commented on Lib/test/test_ctypes/test_dllist.py:
```diff
@@ -0,0 +1,58 @@
+import test.support
+import unittest
+from ctypes import CDLL
+from ctypes.util import dllist
```
Won't this fail with an `ImportError` on the platforms that don't support it? Ideally, we should skip the test with a `unittest.skipUnless(hasattr(ctypes.util, "dllist"))`, or by raising a `SkipTest` at the top-level if it fails.

ZeroIntensity commented on Lib/ctypes/util.py:
```diff
@@ -67,6 +67,64 @@ def find_library(name):
+            if n <= len(modules):
+                return modules[:n]
+
+    def dllist():
```
I think adding a docstring could be helpful. If you want to shorten things a bit, you could add something like this to the bottom of the file:

```py
with contextlib.suppress(NameError):
     dllist.__doc__ = "your silly doc"
```

It's hacky, but ctypes isn't exactly known for it's beautiful code.

WardBrian replied:
I added short (one-liner) docstrings. I went with duplicating it at each of the 3 definitions rather than anything fancy -- I could imagine wanting to expand the docstring to say what system API is used for each version, for example.

ZeroIntensity commented on Lib/ctypes/util.py:
```diff
@@ -67,6 +67,64 @@ def find_library(name):
                 return fname
         return None
 
+    import ctypes
```
I'm *slightly* worried about circular imports ruining our day. I don't think there's anything we can do about that though, I'm just bringing it to attention.

WardBrian replied:
It's certainly a possible problem in the future, but currently it seems to stay out of tying itself it knots

ZeroIntensity commented on Lib/ctypes/util.py:
```diff
@@ -67,6 +67,64 @@ def find_library(name):
+    import ctypes
+    from ctypes import wintypes
+
+    # https://learn.microsoft.com/windows/win32/api/psapi/nf-psapi-enumprocessmodules
```
What do these URLs mean? Could you add a sentence to the comment for clarification?

WardBrian replied:
Each of them is a link to the documentation of the underlying system APIs, mostly for myself and reviewers -- I'm happy to remove them at this point, or add a description as you suggest. I would be slightly concerned about them rotting over time

ZeroIntensity replied:
I think it's good to keep the URLs.

ZeroIntensity commented on Misc/NEWS.d/next/Library/2024-08-12-11-58-15.gh-issue-119349.-xTnHl.rst:
```diff
@@ -0,0 +1,2 @@
+Added the :func:`ctypes.util.dllist` function to list the loaded shared
```
For whatever reason, we do try to use present tense in the changelog. I'm not sure why, but this is the style:
```suggestion
Add the :func:`ctypes.util.dllist` function to list the loaded shared
```

ZeroIntensity commented on Lib/test/test_ctypes/test_dllist.py:
```diff
@@ -0,0 +1,58 @@
+
+            self.assertGreater(dlls2, dlls1)
+            self.assertTrue(any("_ctypes_test" in dll for dll in dlls2))
+
```
Too many newlines:
```suggestion
```

ZeroIntensity commented on Lib/test/test_ctypes/test_dllist.py:
```diff
@@ -0,0 +1,58 @@
+
+        if dlls is not None:
+            if any("_ctypes_test" in dll for dll in dlls):
+                self.skipTest("Test library is already loaded")
```
I'm confused, why is this a skip case?

WardBrian replied:
This test is to check the interaction between dllist and ctypes.CDLL, namely that importing a new shared library updates the list. So, we need to have a shared object which is _not_ already in the list. 

Depending on the specifics of the test runner, if another test was ran before this one by the same process it could have already imported `_ctypes_test`, which would cause this test to spuriously fail

I added a comment to clarify

ZeroIntensity commented on Doc/library/ctypes.rst:
```diff
@@ -1406,6 +1406,19 @@ the shared library name at development time, and hardcode that into the wrapper
 
+.. _ctypes-listing-loaded-shared-libraries:
+
+Listing loaded shared libraries
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+When writing code that relies on code loaded from shared libraries, it can be
+useful to know which shared libraries have already been loaded into the current
+process.
+
+The :mod:`!ctypes.util` module provides the :func:`~ctypes.util.dllist` function,
+which calls the different APIs provided by the various platforms to help determine
+which shared libraries have already been loaded into the current process.
```
Tiny little nitpick: we could add an example here.

WardBrian replied:
I added an example and a slightly longer description to give it context

ZeroIntensity reviewed (commented):
Pretty much LGTM, apart from my many nitpicks. Looking forward to seeing this merged!

WardBrian committed:
test_dllist: check for one of several known libraries

This should cover all tested systems, including Andriod which statically links libffi

WardBrian committed:
test updates per review

WardBrian committed:
Various documentation updates per reviews

ZeroIntensity reviewed (approved):
LGTM! Thanks for doing this and for your patience with reviews, @WardBrian
